tests/unit/s2n_client_hello_retry_test.c

```diff
@@ -0,0 +1,748 @@
+    }
+
+    /* For HelloRetryRequests, test that s2n_client_key_share_extension.recv can read and parse 
+     * the result of s2n_client_key_share_extension.send */
+    {
+        struct s2n_connection *client_conn;
+        struct s2n_connection *server_conn;
+        struct s2n_stuffer key_share_extension;
+
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&key_share_extension, 0));
+
+        /* Setup the client to have received a HelloRetryRequest */
+        memcpy_check(client_conn->secure.server_random, hello_retry_request_random_test_buffer, S2N_TLS_RANDOM_DATA_LEN);
+        client_conn->server_protocol_version = S2N_TLS13;
+        client_conn->client_protocol_version = S2N_TLS13;
+        client_conn->actual_protocol_version = S2N_TLS13;
+        client_conn->handshake.handshake_type = NEGOTIATED | FULL_HANDSHAKE | HELLO_RETRY_REQUEST;
+        client_conn->handshake.message_number = 1;
+        client_conn->secure.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+
+        /* During HRR, A key_share list with a single key_share entry, 
+         * corresponding to the server negotiated curve is sent by the client */
+        EXPECT_SUCCESS(s2n_client_key_share_extension.send(client_conn, &key_share_extension));
+
+        server_conn->secure.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+        server_conn->server_protocol_version = S2N_TLS13;
+        server_conn->client_protocol_version = S2N_TLS13;
+        server_conn->actual_protocol_version = S2N_TLS13;
+
+        EXPECT_SUCCESS(s2n_client_key_share_extension.recv(server_conn, &key_share_extension));
+        /* should read all data */
+        EXPECT_EQUAL(s2n_stuffer_data_available(&key_share_extension), 0);
+
+        const struct s2n_ecc_preferences *ecc_pref = NULL;
+        EXPECT_SUCCESS(s2n_connection_get_ecc_preferences(server_conn, &ecc_pref));
+        EXPECT_NOT_NULL(ecc_pref);
+
+        for (int i = 0; i < ecc_pref->count; i++) {
+            struct s2n_ecc_evp_params *ecc_evp_params = &server_conn->secure.client_ecc_evp_params[i];
+            if (ecc_evp_params->negotiated_curve == server_conn->secure.server_ecc_evp_params.negotiated_curve) {
+                EXPECT_NOT_NULL(ecc_evp_params->negotiated_curve);
+                EXPECT_NOT_NULL(ecc_evp_params->evp_pkey);
+            } else {
+                EXPECT_NULL(ecc_evp_params->negotiated_curve);
+                EXPECT_NULL(ecc_evp_params->evp_pkey);
+            }
+        }
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&key_share_extension));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+    /* Test that s2n_client_key_share_extension.send handles HelloRetryRequests correctly,
```

Comment:
Most of these tests are labeled "Test that X handles HelloRetryRequests correctly". Can you be more specific about what "correctly" means?

Reply:
Correctly refers to they are able to process a HelloRetryRequest in the manner described in the RFC. In the case of s2n_client_key_share_extension.send, it would be to send a list of keyshares containing a single keyshare corresponding to the server selected curve which is mentioned in the s2n_client_key_share_extension.send function.

Reply:
Can you name the tests to indicate that then?